### PR TITLE
Transaction Reversal Implementation 

### PR DESCRIPTION
**Pull request by itsyaasir**

### CODE REVIEW

File: src/services/transaction_reversal.rs [hunk 1]
```diff
@@ -15,17 +20,200 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+        );
+
+        let credentials = self.client.gen_security_credentials()?;
+
+        let payload = TransactionReversalPayload {
+            initiator: self.initiator,
+            security_credentials: &credentials,
+            command_id: self
+                .command_id
+                .clone()
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:186]: Why do we need to clone `command_id`?

File: src/services/transaction_reversal.rs [hunk 2]
```diff
@@ -15,17 +20,200 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+        self
+    }
+
+    /// Adds an `amount` to the request
+    ///
+    /// This is a required field
+    pub fn amount(mut self, amount: f64) -> Self {
+        self.amount = Some(amount);
+        self
+    }
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:152]: Could you make `amount` generic enough to accept whole numbers and decimals?

```suggestion
    pub fn amount<Number: Into<f64>>(mut self, amount: Number) -> Self {
        self.amount = Some(amount.into());
        self
    }
```

File: tests/mpesa-rust/transaction_reversal_test.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        .result_url("https://testdomain.com/ok")
+        .timeout_url("https://testdomain.com/err")
+        .transaction_id("OEI2AK4Q16")
+        .amount(1.0)
+        .receiver_party("600983")
+        .remarks("wrong recipient")
+        .send()
+        .await;
+
+    assert!(response.is_ok())
```
Comment (collinsmuriuki) [inline tests/mpesa-rust/transaction_reversal_test.rs:18]: Quick favour: could you enable ci jobs to run on `pull_request` too in `.github/workflows/general.yml` as part of your PR?

**Review by collinsmuriuki — CHANGES_REQUESTED** [review]

File: tests/mpesa-rust/transaction_reversal_test.rs (hunk 3 above)
Comment (collinsmuriuki) [reply tests/mpesa-rust/transaction_reversal_test.rs:18]: wait just checked that `pull_request` is included in the list, will investigate why ci jobs are not running

Comment (collinsmuriuki) [reply tests/mpesa-rust/transaction_reversal_test.rs:18]: need to approve the PR before the jobs run, will change this.

File: src/services/transaction_reversal.rs [hunk 4]
```diff
@@ -15,17 +20,200 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+
+    /// Adds `QueueTimeoutUrl` and `ResultUrl`. This is a required field
+    ///
+    /// # Error
+    /// If either `QueueTimeoutUrl` and `ResultUrl` is invalid or not provided
+    pub fn timeout_url(mut self, timeout_url: &'mpesa str) -> Self {
+        self.timeout_url = Some(timeout_url);
+        self
+    }
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:131]:
```suggestion
    }
    
```

File: src/services/transaction_reversal.rs [hunk 5]
```diff
@@ -15,17 +20,200 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+    /// Comments that are sent along with the transaction.
+    ///
+    /// This is required field
+    pub fn remarks(mut self, remarks: &'mpesa str) -> Self {
+        self.remarks = Some(remarks);
+        self
+    }
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:138]:
```suggestion
    }
    
```

**Review by collinsmuriuki — CHANGES_REQUESTED** [review]

File: src/services/transaction_reversal.rs [hunk 6]
```diff
@@ -15,17 +20,200 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+pub struct TransactionReversalResponse {
+    #[serde(rename(deserialize = "ConversationID"))]
+    conversation_id: String,
+    #[serde(rename(deserialize = "OriginatorConversationID"))]
+    originator_conversation_id: String,
+    #[serde(rename(deserialize = "ResponseDescription"))]
+    response_description: String,
+}
+
+/// Creates new `TransactionReversalBuilder`
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:46]: I think this comment should be on the builder's `new` method

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:46]: True

File: src/services/transaction_reversal.rs (hunk 2 above)
Comment (itsyaasir) [reply src/services/transaction_reversal.rs:154]: Yep

File: src/services/transaction_reversal.rs (hunk 1 above)
Comment (itsyaasir) [reply src/services/transaction_reversal.rs:188]: oversight, let me get rid of that

**Review by collinsmuriuki — APPROVED** [review]

**Comment by collinsmuriuki** [comment]
@itsyaasir you will probably have to add your `CLIENT_ID` and `CLIENT_SECRET` [here](https://github.com/itsyaasir/mpesa-rust/settings/secrets/actions) in order for tests to pass. This is an inconveniencing step I am working on removing with implementation of black box testing, so that we don't have to make actual api calls in our tests

File: src/services/transaction_reversal.rs [hunk 7]
```diff
@@ -174,7 +174,7 @@ impl<'mpesa, Env: ApiEnvironment> TransactionReversalBuilder<'mpesa, Env> {
     /// Returns a `MpesaError` on failure.
     pub async fn send(self) -> MpesaResult<TransactionReversalResponse> {
         let url = format!(
-            "{}/reversal/v1/request",
+            "{}/mpesa/reversal/v1/request",
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:177]: Good catch!

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:177]: But the test still failing :/ due to invalid `ReceiverIdentifierType`, I used the default `11` from their site but didn't work.

Comment (collinsmuriuki) [reply src/services/transaction_reversal.rs:177]: I think `"4"` should be the valid `ReceiverIdentifierType`, which should be `IdentifierTypes::ShortCode`

File: src/services/transaction_reversal.rs [hunk 8]
```diff
@@ -15,17 +20,199 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+    /// This is required field
+    pub fn receiver_party(mut self, receiver_party: &'mpesa str) -> Self {
+        self.receiver_party = Some(receiver_party);
+        self
+    }
+
+    /// Type of organization receiving the transaction
+    ///
+    /// This is required field
+    pub fn receiver_identifier_type(mut self, receiver_identifier_type: &'mpesa str) -> Self {
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:110]: Reciever identifier types are represented as an enum `IdentifierType` with the default for this api being `IdentifierType::ShortCode`

```suggestion
    pub fn receiver_identifier_type(mut self, receiver_identifier_type: IdentifierType) -> Self {
```

**Review by collinsmuriuki — CHANGES_REQUESTED** [review]

File: src/services/transaction_reversal.rs (hunk 8 above)
Comment (itsyaasir) [reply src/services/transaction_reversal.rs:110]: Let me work on the fix

File: src/services/transaction_reversal.rs (hunk 7 above)
Comment (collinsmuriuki) [reply src/services/transaction_reversal.rs:177]: I just realized you were not using the `IdentifierTypes` for the `identifier_type` parameter, I have added it with this suggestion https://github.com/collinsmuriuki/mpesa-rust/pull/44#discussion_r1027111429

Make sure you also update the tests

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:177]: Doesn't seem to work :/

Comment (collinsmuriuki) [reply src/services/transaction_reversal.rs:177]: Still getting the same error?

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:177]: Yes, even on the saf API

Comment (collinsmuriuki) [reply src/services/transaction_reversal.rs:177]: Could you share a snippet of the TransactionReversalBuilder's `send` method you are using to test this?

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:177]:
```
pub async fn send(self) -> MpesaResult<TransactionReversalResponse> {
        let url = format!(
            "{}/mpesa/reversal/v1/request",
            self.client.environment().base_url()
        );

        let credentials = self.client.gen_security_credentials()?;

        let payload = TransactionReversalPayload {
            initiator: self.initiator,
            security_credentials: &credentials,
            command_id: self.command_id.unwrap_or(CommandId::TransactionReversal),
            transaction_id: self
                .transaction_id
                .ok_or(MpesaError::Message("transaction_id is required field"))?,
            receiver_party: self
                .receiver_party
                .ok_or(MpesaError::Message("receiver_party is required field"))?,
            receiver_identifier_type: self.receiver_identifier_type,
            result_url: self.result_url,
            timeout_url: self.timeout_url,
            remarks: self.remarks,
            occasion: self.occasion,
            amount: self
                .amount
                .ok_or(MpesaError::Message("amount is required field"))?,
        };

        let response = self
            .client
            .http_client
            .post(&url)
            .bearer_auth(self.client.auth().await?)
            .json(&payload)
            .send()
            .await?;

        if !response.status().is_success() {
            let value = response.json().await?;
            return Err(MpesaError::MpesaTransactionReversalError(value));
        };

        let response = response.json::<_>().await?;
        Ok(response)
    }
```

**Comment by itsyaasir** [comment]
All tests are passing, we can merge this in.

File: src/services/transaction_reversal.rs [hunk 9]
```diff
@@ -14,18 +20,202 @@ pub struct TransactionReversalPayload<'mpesa> {
     transaction_id: &'mpesa str,
     #[serde(rename(serialize = "ReceiverParty"))]
     receiver_party: &'mpesa str,
-    #[serde(rename(serialize = "ReceiverIdentifierType"))]
-    receiver_identifier_type: &'mpesa str,
+    #[serde(rename(serialize = "RecieverIdentifierType"))]
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:23]: Wow, they don't know English? 😂

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:23]: Saidia wao bwanaa, typo kaa izi inazungusha watu

File: src/services/transaction_reversal.rs [hunk 10]
```diff
@@ -14,18 +20,202 @@ pub struct TransactionReversalPayload<'mpesa> {
...
+        let payload = TransactionReversalPayload {
+            initiator: self.initiator,
+            security_credentials: &credentials,
+            command_id: self.command_id.unwrap_or(CommandId::TransactionReversal),
+            transaction_id: self
+                .transaction_id
+                .ok_or(MpesaError::Message("transaction_id is required field"))?,
+            receiver_party: self
+                .receiver_party
+                .ok_or(MpesaError::Message("receiver_party is required field"))?,
```
Comment (collinsmuriuki) [inline src/services/transaction_reversal.rs:193]: Good addition, this is needed for other builders, can be done in another PR

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:193]: Yes, will take up that task.

Comment (itsyaasir) [reply src/services/transaction_reversal.rs:193]: I think we should add a type for this error, so we don't have to type same message everytime.

**Comment by itsyaasir** [comment]
It failed again, but I think I know the reason.

> Secrets are not passed to workflows that are triggered by a pull request from a fork

**Pull request merged**